src/timely-util/src/reclock.rs

```diff
@@ -0,0 +1,609 @@
+    let (mut output, reclocked) = builder.new_output();
+
+    builder.build(move |caps| {
+        let [cap]: [_; 1] = caps.try_into().expect("one output");
+
+        // Remap updates beyond the upper
+        let mut accepted_bindings = Vec::new();
+        let mut replayer = BindingReplay::new(cap);
+
+        // The upper frontier of updates in source_batches
+        let mut source_upper = MutableAntichain::new_bottom(Timestamp::minimum());
+        // Batches of data waiting to be reclocked
+        let mut source_batches = VecDeque::new();
+
+        let mut vector = Vec::new();
+        let mut container = Vec::new();
+        move |frontiers| {
+            // Accept new bindings
+            while let Some((_, data)) = remap_input.next() {
+                data.swap(&mut vector);
+                accepted_bindings.append(&mut vector);
+            }
+            // Extract ready batches from accepted bindings and update the replayer
+            let upper = frontiers[0].frontier();
+            if PartialOrder::less_than(&replayer.upper(), &upper) {
+                accepted_bindings.sort_unstable_by(|a, b| a.1.cmp(&b.1));
+                // The times are totally ordered so we can binary search to find the prefix that is
+                // not beyond the upper and extract it into a batch.
+                let idx = accepted_bindings.partition_point(|(_, t, _)| !upper.less_equal(t));
+                replayer.append_batch(accepted_bindings.drain(0..idx), upper.to_owned());
+            }
+
+            // Accept new data
+            while let Some(event) = puller.pull() {
+                match event.as_mut() {
+                    Event::Progress(changes) => {
+                        source_upper.update_iter(changes.drain(..));
+                    }
+                    Event::Messages(_, data) => {
+                        data.sort_unstable_by(|a: &(D1, T, R), b| a.1.cmp(&b.1));
+                        let batch = Batch::from_iter(data.drain(..));
+                        source_batches.push_back(batch);
+                    }
+                }
+            }
+
+            // Reclock accepted data
+            let mut output = output.activate();
+            while let Some((frontier, cap)) = replayer.active_binding() {
+                let into_t = cap.time();
+                source_batches.retain_mut(|batch| {
+                    batch.chains.retain_mut(|chain| {
+                        // A chain is a set of mutually comparable elements so we can binary search
+                        // to find the prefix that is not beyond the frontier and extract it.
+                        let idx = chain.partition_point(|(_, t, _)| !frontier.less_equal(t));
+                        let data = chain
+                            .drain(0..idx)
+                            .map(|(d, t, r)| (logic(d, t), into_t.clone(), r));
+                        container.extend(data);
```

Comment:
I think you can grab a session earlier and use the regular `give` API here. This avoids the risk of `container` growing larger than a regular batch size.

Reply:
done